official/coin_build/man/man3/ScXMLAnchorElt.3
.TH "ScXMLAnchorElt" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ScXMLAnchorElt \- implements the <anchor> SCXML element\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/scxml/ScXMLAnchorElt\&.h>\fP
.PP
Inherits \fBScXMLElt\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "virtual void \fBsetTypeAttribute\fP (const char *type)"
.br
.ti -1c
.RI "const char * \fBgetTypeAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetSnapshotAttribute\fP (const char *snapshot)"
.br
.ti -1c
.RI "const char * \fBgetSnapshotAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual SbBool \fBhandleXMLAttributes\fP (void)"
.br
.ti -1c
.RI "virtual void \fBcopyContents\fP (const \fBScXMLElt\fP *rhs)"
.br
.ti -1c
.RI "virtual const \fBScXMLElt\fP * \fBsearch\fP (const char *attrname, const char *attrvalue) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBScXMLEltReader\fP * \fBgetElementReader\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void * \fBcreateInstance\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBcleanClass\fP (void)"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "char * \fBtype\fP"
.br
.ti -1c
.RI "char * \fBsnapshot\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
implements the <anchor> SCXML element\&. 

The <anchor> element is not supported yet other than as a dummy state\&.
.PP
\fBSince:\fP
.RS 4
Coin 3\&.0 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP ScXMLAnchorElt::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. 
.PP
\fBSee also:\fP
.RS 4
\fBSoBase::getTypeId() const\fP 
.RE
.PP

.PP
Implements \fBScXMLElt\fP\&.
.SS "SbBool ScXMLAnchorElt::handleXMLAttributes (void)\fC [virtual]\fP"
This method is called when the file reader has set all the XML attributes on the object, and wants the object to handle them and report if the values were ok or if a read error should be produced\&.
.PP
\fBReturns:\fP
.RS 4
TRUE if the attributes are ok, and FALSE on error\&.
.RE
.PP
This base class implementation does nothing and just returns TRUE\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.
.SS "const \fBScXMLElt\fP * ScXMLAnchorElt::search (const char * attrname, const char * attrvalue) const\fC [virtual]\fP"
This method searches the SCXML structure for an element with the given attributevalue for the given attribute\&.
.PP
Returns NULL if nothing was found\&.
.PP
This function needs to be reimplemented to traverse child elements\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/scxml/ScXMLAnchorElt.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

#include <Inventor/scxml/ScXMLAnchorElt.h>

/*!
  \class ScXMLAnchorElt ScXMLAnchorElt.h Inventor/scxml/ScXMLAnchorElt.h
  \brief implements the &lt;anchor&gt; SCXML element.

  The &lt;anchor&gt; element is not supported yet other than as a
  dummy state.

  \since Coin 3.0
  \ingroup scxml
*/

#include <cassert>
#include <cstring>

#include <Inventor/errors/SoDebugError.h>
#include <Inventor/C/XML/element.h>
#include <Inventor/scxml/ScXML.h>

#include "scxml/ScXMLCommonP.h"
#include "SbBasicP.h"

// *************************************************************************

class ScXMLAnchorEltReader : public ScXMLEltReader {
public:
  ScXMLAnchorEltReader(void);
  virtual ScXMLElt * read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * doc, ScXMLStateMachine * sm);
};

ScXMLAnchorEltReader::ScXMLAnchorEltReader(void)
: ScXMLEltReader("anchor")
{
}

ScXMLElt *
ScXMLAnchorEltReader::read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * COIN_UNUSED_ARG(doc), ScXMLStateMachine * COIN_UNUSED_ARG(sm))
{
  assert(container && xmlelt);
  ScXMLAnchorElt * anchor = new ScXMLAnchorElt;
  anchor->setContainer(container);
  this->setXMLAttributes(anchor, xmlelt);

  // handle XML attributes
  if (unlikely(!anchor->handleXMLAttributes())) {
    delete anchor;
    return nullptr;
  }

  const int numchildren = cc_xml_elt_get_num_children(xmlelt);
  for (int c = 0; c < numchildren; ++c) {
    cc_xml_elt * element = cc_xml_elt_get_child(xmlelt, c);
    const char * elementtype = cc_xml_elt_get_type(element);

    if (strcmp(elementtype, COIN_XML_CDATA_TYPE) == 0) {
      // ignore CDATA
      continue;
    }

    SoDebugError::post("ScXMLAnchorEltReader::read",
                       "unexpected XML element '<%s>' found in <anchor>",
                       elementtype);
    delete anchor;
    return nullptr;
  }

  return anchor;
}

// *************************************************************************

class ScXMLAnchorElt::PImpl {
public:
};

#define PRIVATE

SCXML_ELEMENT_SOURCE(ScXMLAnchorElt);

void
ScXMLAnchorElt::initClass(void)
{
  SCXML_OBJECT_INIT_CLASS(ScXMLAnchorElt, ScXMLElt, "ScXMLElt");
  SCXML_ELEMENT_REGISTER_READER(ScXMLAnchorElt, "anchor", ScXMLAnchorEltReader);
}

void
ScXMLAnchorElt::cleanClass(void)
{
  SCXML_ELEMENT_UNREGISTER_READER(ScXMLAnchorElt);
  ScXMLAnchorElt::classTypeId = SoType::badType();
}

ScXMLAnchorElt::ScXMLAnchorElt(void)
: type(nullptr),
  snapshot(nullptr)
{
}

ScXMLAnchorElt::~ScXMLAnchorElt(void)
{
  this->setTypeAttribute(nullptr);
  this->setSnapshotAttribute(nullptr);
}

void
ScXMLAnchorElt::setTypeAttribute(const char * typestr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->type, "type", typestr);
}

void
ScXMLAnchorElt::setSnapshotAttribute(const char * snapshotstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->snapshot, "snapshot", snapshotstr);
}

bool
ScXMLAnchorElt::handleXMLAttributes(void)
{
  if (!inherited::handleXMLAttributes()) return false;

  this->setTypeAttribute(this->getXMLAttribute("type"));
  this->setSnapshotAttribute(this->getXMLAttribute("snapshot"));

  if (this->type == nullptr) {
    return false;
  }

  return true;
}

void
ScXMLAnchorElt::copyContents(const ScXMLElt * rhs)
{
  inherited::copyContents(rhs);
  const ScXMLAnchorElt * orig = coin_assert_cast<const ScXMLAnchorElt *>(rhs);
  this->setTypeAttribute(orig->getTypeAttribute());
  this->setSnapshotAttribute(orig->getSnapshotAttribute());
}

const ScXMLElt *
ScXMLAnchorElt::search(const char * attrname, const char * attrvalue) const
{
  const ScXMLElt * hit = inherited::search(attrname, attrvalue);
  if (hit) {
    return hit;
  }
  if (strcmp(attrname, "type") == 0) {
    if (this->type && strcmp(attrvalue, this->type) == 0) {
      return this;
    }
  }
  else if (strcmp(attrname, "snapshot") == 0) {
    if (this->snapshot && strcmp(attrvalue, this->snapshot) == 0) {
      return this;
    }
  }
  return nullptr;
}

#undef PRIVATE
